flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtaskAllExecutionAttemptsDetailsHandler.java

```diff
@@ -0,0 +1,90 @@
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.executiongraph.AccessExecutionVertex;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.rest.handler.HandlerRequest;
+import org.apache.flink.runtime.rest.handler.RestHandlerException;
+import org.apache.flink.runtime.rest.handler.legacy.ExecutionGraphCache;
+import org.apache.flink.runtime.rest.handler.legacy.metrics.MetricFetcher;
+import org.apache.flink.runtime.rest.messages.EmptyRequestBody;
+import org.apache.flink.runtime.rest.messages.JobIDPathParameter;
+import org.apache.flink.runtime.rest.messages.JobVertexIdPathParameter;
+import org.apache.flink.runtime.rest.messages.MessageHeaders;
+import org.apache.flink.runtime.rest.messages.job.SubtaskAllExecutionAttemptsDetailsInfo;
+import org.apache.flink.runtime.rest.messages.job.SubtaskExecutionAttemptDetailsInfo;
+import org.apache.flink.runtime.rest.messages.job.SubtaskMessageParameters;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.util.Preconditions;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+/**
+ * Handler which returns the details of all execution attempts of a subtask.
+ */
+public class SubtaskAllExecutionAttemptsDetailsHandler
+	extends AbstractSubtaskHandler<SubtaskAllExecutionAttemptsDetailsInfo, SubtaskMessageParameters> {
+
+	private final MetricFetcher metricFetcher;
+
+	public SubtaskAllExecutionAttemptsDetailsHandler(
+		GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+		Time timeout,
+		Map<String, String> responseHeaders,
+		MessageHeaders<EmptyRequestBody, SubtaskAllExecutionAttemptsDetailsInfo, SubtaskMessageParameters> messageHeaders,
+		ExecutionGraphCache executionGraphCache,
+		Executor executor,
+		MetricFetcher metricFetcher) {
+		super(leaderRetriever, timeout, responseHeaders, messageHeaders, executionGraphCache, executor);
+
+		this.metricFetcher = Preconditions.checkNotNull(metricFetcher);
+	}
+
+	@Override
+	protected SubtaskAllExecutionAttemptsDetailsInfo handleRequest(
+			HandlerRequest<EmptyRequestBody, SubtaskMessageParameters> request,
+			AccessExecutionVertex executionVertex) throws RestHandlerException {
+		final JobID jobID = request.getPathParameter(JobIDPathParameter.class);
+		final JobVertexID jobVertexID = request.getPathParameter(JobVertexIdPathParameter.class);
+
+		return createSubtaskExecutionAttemptsDetailsInfo(executionVertex, jobID, jobVertexID);
+	}
+
+	protected SubtaskAllExecutionAttemptsDetailsInfo createSubtaskExecutionAttemptsDetailsInfo(
+		AccessExecutionVertex executionVertex,
+		JobID jobID,
+		JobVertexID jobVertexID) {
+		List<SubtaskExecutionAttemptDetailsInfo> allAttempts = new ArrayList<>();
+		executionVertex.getPriorExecutionAttempts().forEach(execution -> {
+			if (execution != null) {
```

Comment:
Can you describe when `execution` can be null in the prior execution attempts list?

Reply:
In [JIAR](https://issues.apache.org/jira/browse/FLINK-12247) has talked about it.

Reply:
Can we filter nulls in `executionVertex.getPriorExecutionAttempts()`? It's a strange contract to return a list with nulls.